src/middlewared/middlewared/plugins/zfs_/utils.py

```diff
@@ -299,3 +308,53 @@ def iter_datasets(out, datasets_in, batch_ops):
             logger.warning('Failed to update cached snapshot counts', exc_info=True)
 
     return out
+
+
+def paths_to_datasets_impl(
+    paths: list[str],
+    mntinfo: dict | None = None
+) -> dict | dict[str, str]:
+    """
+    Convert `paths` to a dictionary of ZFS dataset names. This
+    performs lookup through mountinfo.
+
+    Anticipated error conditions are that paths are not
+    on ZFS or if the boot pool underlies the path. In
+    addition to this, all the normal exceptions that
+    can be raised by a failed call to os.stat() are
+    possible.
+    """
+    rv = dict()
+    if mntinfo is None:
+        mntinfo = getmntinfo()
+
+    for path in paths:
+        rv[path] = path_to_dataset_impl(path, mntinfo)
```

Comment:
We should probably have exception handling here for some path which if we do get an error, we will miss out on all the data we did want i.e so we can probably have null for such paths